### ISSUE TITLE
Feature request: Allow bindings to return values

### ISSUE THREAD

**Issue by MikkelSchubert** [typed-argparse/typed-argparse#58]
It would be handy if `run()` would return the result of the selected binding. Currently it always returns None. This would for example allow sub-commands to return an exit code rather than each having to individually call `exit`:

```
def run_foo(args: FooArgs) -> int:
    return 1  # failed

def run_bar(args: BarArgs) -> int:
    return 0  # okay

def main(argv: List[str]) -> int:
    return tap.Parser(...).bind(run_foo, run_bar).run(argv)

if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
```

**Comment by bluenote10** [comment]
Yes indeed, that sounds very reasonable. The interesting question will become if we can also make it fully type-safe so that the return type are guaranteed to be consistent and are properly forwarded. I.e., ideally `reveal_type(tap.Parser(...).bind(run_foo, run_bar).run(argv))` in your example should reveal `int`.

**Comment by MikkelSchubert** [comment]
That seems doable, since bind returns an App rather than self.

I've made an attempt at it (see attached patch), and it seems to work as expected.
But I'm also not very confident in my understanding of type annotations so *caveat emptor*.

[typed_return.txt](https://github.com/typed-argparse/typed-argparse/files/12357129/typed_return.txt)

**Comment by bluenote10** [comment]
Nice! Yes that looks pretty much like what I would have tried first. Will have to experiment a bit with it, if everything works fine (e.g. if the use of explicit `Binding` also works, and what happens when mixing things etc.).

**Comment by bluenote10** [comment]
I was looking into this a bit, and opened a PR container the suggested changes. However it looks like this doesn't really type check, does it? For instance in the following basic example, mypy seems to infer the app has type `App[Any]` indicating that it loses the type information.

```py
class FooArgs(TypedArgs):
    ...

class BarArgs(TypedArgs):
    ...

parser = Parser(
    SubParserGroup(
        SubParser("foo", FooArgs),
        SubParser("bar", BarArgs),
    )
)

def foo(foo_args: FooArgs) -> int:
    return 42

def bar(bar_args: BarArgs) -> str:
    return "foo"

app = parser.bind(Binding(FooArgs, foo), Binding(BarArgs, bar))
reveal_type(app) # say App[Any] not App[int | str] as hoped.
```

It looks to me that mypy is missing support to properly take the union of return types in general. I've opened an issue for that on mypy: https://github.com/python/mypy/issues/15968

**Comment by bluenote10** [comment]
> However, I think you also need to change ... to ...

Good catch, I had missed that occurrence of `None` in my search/replace action.

I've pushed the change to the PR, including more type checks which show the current `# type: ignore`s needed to make it work.

The case of heterogeneous return types doesn't look good in mypy. Pyright does a better job here -- which is basically a result of the "join v union" issue mentioned in the linked mypy discussion.

But perhaps the case of heterogeneous return types is a bit exotic. I'd assume that it is much more common that _all_ bound methods should return the same types (typically `None` or `int`). So it would be okay-ish if accidentally adding a method that has a different return type results in `App[object]`, because one would actually see the problem when doing e.g. `sys.exit(return_code)` and `return_code` has degraded into `object`. Only `App[Any]` is worrying.

So I would mainly focus on getting the _homogeneous_ case right. Here it seems that 1 out of the 4 basic cases (eager vs lazy; using `Binding` vs plain functions) does not work yet: The case of lazy + plain functions case. Such code currently type checks, so it would be unfortunate if this would break. I'll have to take another look if there is something we could do about it.
